File: internal/agentrun/adapter.go
package agentrun

import (
	"errors"
	"regexp"
	"sort"
	"strings"
	"time"
)

const ProviderEventModelVersion uint = 1

var stableProviderEventName = regexp.MustCompile(`^[A-Za-z][A-Za-z0-9._-]*$`)

// ProviderEvent is the closed hook/wrapper input shared by provider adapters.
// Provider identity is intentionally absent: the authenticated Run supplies
// it. Adapters never accept native raw JSON, prompts, messages, tool payloads,
// command arguments, output, transcripts, environment, or credentials.
type ProviderEvent struct {
	ModelVersion uint      `json:"modelVersion"`
	ID           string    `json:"id"`
	Sequence     uint64    `json:"sequence"`
	Type         string    `json:"type"`
	Category     EventKind `json:"category,omitempty"`
	Subject      string    `json:"subject,omitempty"`
	Paths        []string  `json:"paths,omitempty"`
	CommitSHA    string    `json:"commitSha,omitempty"`
	ExitCode     *int      `json:"exitCode,omitempty"`
	ErrorClass   string    `json:"errorClass,omitempty"`
	Summary      string    `json:"summary,omitempty"`
	OccurredAt   time.Time `json:"occurredAt,omitempty"`
}

type adapterMapping struct {
	kind         EventKind
	phase        EventPhase
	outcome      EventOutcome
	errorClass   string
	notification EventNotificationKind
}

var providerEventAliases = map[string]map[string]adapterMapping{
	"codex": {
		"sessionstart":       {kind: EventLifecycle, phase: EventStarted},
		"turn.started":       {kind: EventLifecycle, phase: EventProgress},
		"turn.completed":     {kind: EventLifecycle, phase: EventWaiting},
		"turn.failed":        {kind: EventError, phase: EventFailed, outcome: EventUnsuccessful, notification: NotificationFailure},
		"pretooluse":         {kind: EventTool, phase: EventStarted},
		"posttooluse":        {kind: EventTool, phase: EventCompleted, outcome: EventSucceeded},
		"posttoolusefailure": {kind: EventTool, phase: EventFailed, outcome: EventUnsuccessful},
		"permissionrequest":  {kind: EventLifecycle, phase: EventWaiting, notification: NotificationApprovalRequested},
		"question":           {kind: EventLifecycle, phase: EventWaiting, notification: NotificationQuestion},
		"stop":               {kind: EventLifecycle, phase: EventWaiting},
		"sessionend":         {kind: EventLifecycle, phase: EventCompleted, outcome: EventSucceeded, notification: NotificationCompletion},
	},
	"claude": {
		"sessionstart":       {kind: EventLifecycle, phase: EventStarted},
		"pretooluse":         {kind: EventTool, phase: EventStarted},
		"posttooluse":        {kind: EventTool, phase: EventCompleted, outcome: EventSucceeded},
		"posttoolusefailure": {kind: EventTool, phase: EventFailed, outcome: EventUnsuccessful},
		"permissionrequest":  {kind: EventLifecycle, phase: EventWaiting, notification: NotificationApprovalRequested},
		"question":           {kind: EventLifecycle, phase: EventWaiting, notification: NotificationQuestion},
		"stop":               {kind: EventLifecycle, phase: EventWaiting},
		"sessionend":         {kind: EventLifecycle, phase: EventCompleted, outcome: EventSucceeded, notification: NotificationCompletion},
	},
	"gemini": {
		"sessionstart":      {kind: EventLifecycle, phase: EventStarted},
		"beforetool":        {kind: EventTool, phase: EventStarted},
		"aftertool":         {kind: EventTool, phase: EventCompleted, outcome: EventSucceeded},
		"notification":      {kind: EventLifecycle, phase: EventProgress},
		"permissionrequest": {kind: EventLifecycle, phase: EventWaiting, notification: NotificationApprovalRequested},
		"question":          {kind: EventLifecycle, phase: EventWaiting, notification: NotificationQuestion},
		"sessionend":        {kind: EventLifecycle, phase: EventCompleted, outcome: EventSucceeded, notification: NotificationCompletion},
	},
	"agy": {
		"session.started":      {kind: EventLifecycle, phase: EventStarted},
		"session.waiting":      {kind: EventLifecycle, phase: EventWaiting},
		"session.completed":    {kind: EventLifecycle, phase: EventCompleted, outcome: EventSucceeded, notification: NotificationCompletion},
		"session.failed":       {kind: EventLifecycle, phase: EventFailed, outcome: EventUnsuccessful, notification: NotificationFailure},
		"session.question":     {kind: EventLifecycle, phase: EventWaiting, notification: NotificationQuestion},
		"permission.requested": {kind: EventLifecycle, phase: EventWaiting, notification: NotificationApprovalRequested},
	},
	"opencode": {
		"session.created":      {kind: EventLifecycle, phase: EventStarted},
		"session.idle":         {kind: EventLifecycle, phase: EventWaiting},
		"session.error":        {kind: EventError, phase: EventFailed, outcome: EventUnsuccessful, errorClass: "session_failure", notification: NotificationFailure},
		"session.completed":    {kind: EventLifecycle, phase: EventCompleted, outcome: EventSucceeded, notification: NotificationCompletion},
		"session.question":     {kind: EventLifecycle, phase: EventWaiting, notification: NotificationQuestion},
		"permission.requested": {kind: EventLifecycle, phase: EventWaiting, notification: NotificationApprovalRequested},
		"tool.execute.before":  {kind: EventTool, phase: EventStarted},
		"tool.execute.after":   {kind: EventTool, phase: EventCompleted, outcome: EventSucceeded},
		"file.edited":          {kind: EventFile, phase: EventCompleted, outcome: EventSucceeded},
		"command.executed":     {kind: EventCommand, phase: EventCompleted, outcome: EventSucceeded},
	},
}

// SupportedEventProviders returns the installed provider-adapter identities
// in deterministic order. Unknown future providers receive only the generic,
// explicit lifecycle fallback.
func SupportedEventProviders() []string {
	providers := make([]string, 0, len(providerEventAliases))
	for provider := range providerEventAliases {
		providers = append(providers, provider)
	}
	sort.Strings(providers)
	return providers
}

// NormalizeProviderEvent maps one allowlisted provider hook event into the
// provider-neutral observation contract. Known providers may also send the
// canonical `<kind>.<phase>` form through an explicit wrapper. Unknown future
// providers are lifecycle-only until an adapter is deliberately added.
func NormalizeProviderEvent(provider string, input ProviderEvent) (EventObservation, error) {
	provider = strings.ToLower(strings.TrimSpace(provider))
	if provider == "" || len(provider) > 64 || !stableProviderEventName.MatchString(provider) {
		return EventObservation{}, errors.New("invalid agent event provider")
	}
	if input.ModelVersion != ProviderEventModelVersion || input.Sequence == 0 {
		return EventObservation{}, errors.New("unsupported provider event contract")
	}
	eventType := strings.TrimSpace(input.Type)
	if eventType == "" || len(eventType) > 64 || !stableProviderEventName.MatchString(eventType) {
		return EventObservation{}, errors.New("invalid provider event type")
	}

	mapping, mapped := providerEventAliases[provider][strings.ToLower(eventType)]
	if provider == "codex" && !mapped {
		mapping, mapped = normalizeCodexItem(eventType, input.Category)
	}
	if !mapped {
		mapping, mapped = canonicalProviderEvent(eventType)
		if mapped {
			if mapping.kind == EventSummary {
				return EventObservation{}, errors.New("provider summaries require an explicit trusted adapter")
			}
			_, knownProvider := providerEventAliases[provider]
			if !knownProvider && mapping.kind != EventLifecycle {
				return EventObservation{}, errors.New("future provider adapter is lifecycle-only")
			}
		}
	}
	if input.Category != "" && !(provider == "codex" && strings.HasPrefix(strings.ToLower(eventType), "item.")) {
		return EventObservation{}, errors.New("provider event category is not allowed for this type")
	}
	if input.ExitCode != nil && *input.ExitCode != 0 &&
		(mapping.kind == EventLifecycle || mapping.kind == EventCommand || mapping.kind == EventTest) {
		mapping.phase = EventFailed
		mapping.outcome = EventUnsuccessful
		if mapping.notification != "" {
			mapping.notification = NotificationFailure
		}
	}
	if mapping.notification != "" {
		return EventObservation{
			ModelVersion: EventModelVersion, SourceID: input.ID,
			SourceSequence: input.Sequence, Kind: mapping.kind, Phase: mapping.phase,
			Outcome: mapping.outcome, ErrorClass: mapping.errorClass,
			OccurredAt: input.OccurredAt, Notification: mapping.notification,
			recognizedNotification: true,
		}, nil
	}
	if !mapped {
		return EventObservation{}, errors.New("unsupported provider event type")
	}
	errorClass := input.ErrorClass
	if errorClass == "" {
		errorClass = mapping.errorClass
	}
	return EventObservation{
		ModelVersion:   EventModelVersion,
		SourceID:       input.ID,
		SourceSequence: input.Sequence,
		Kind:           mapping.kind,
		Phase:          mapping.phase,
		Outcome:        mapping.outcome,
		Subject:        input.Subject,
		Paths:          append([]string{}, input.Paths...),
		CommitSHA:      input.CommitSHA,
		ExitCode:       cloneInt(input.ExitCode),
		ErrorClass:     errorClass,
		Summary:        input.Summary,
		OccurredAt:     input.OccurredAt,
	}, nil
}

func normalizeCodexItem(eventType string, category EventKind) (adapterMapping, bool) {
	var phase EventPhase
	var outcome EventOutcome
	switch strings.ToLower(eventType) {
	case "item.started":
		phase = EventStarted
	case "item.completed":
		phase = EventCompleted
		outcome = EventSucceeded
	case "item.failed":
		phase = EventFailed
		outcome = EventUnsuccessful
	default:
		return adapterMapping{}, false
	}
	if !category.Valid() || category == EventLifecycle || category == EventSummary {
		return adapterMapping{}, false
	}
	return adapterMapping{kind: category, phase: phase, outcome: outcome}, true
}

func canonicalProviderEvent(eventType string) (adapterMapping, bool) {
	kindName, phaseName, found := strings.Cut(strings.ToLower(eventType), ".")
	if !found || strings.Contains(phaseName, ".") {
		return adapterMapping{}, false
	}
	kind := EventKind(kindName)
	phase := EventPhase(phaseName)
	if !kind.Valid() || !phase.Valid() {
		return adapterMapping{}, false
	}
	outcome := EventOutcome("")
	if phase == EventCompleted {
		outcome = EventSucceeded
	} else if phase == EventFailed {
		outcome = EventUnsuccessful
	}
	notification := EventNotificationKind("")
	if kind == EventLifecycle && phase == EventCompleted {
		notification = NotificationCompletion
	} else if kind == EventLifecycle && phase == EventFailed {
		notification = NotificationFailure
	}
	return adapterMapping{
		kind: kind, phase: phase, outcome: outcome, notification: notification,
	}, true
}

File: crates/ptrack-agent/src/adapter.rs
use std::fmt;

use serde::{Deserialize, Serialize};

use crate::{
    EVENT_MODEL_VERSION, EventKind, EventNotificationKind, EventObservation, EventOutcome,
    EventPhase, Timestamp,
};

pub const PROVIDER_EVENT_MODEL_VERSION: u32 = 1;

#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields)]
#[serde(rename_all = "camelCase")]
pub struct ProviderEvent {
    pub model_version: u32,
    pub id: String,
    pub sequence: u64,
    #[serde(rename = "type")]
    pub event_type: String,
    #[serde(default, skip_serializing_if = "is_unset_kind")]
    pub category: EventKind,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub subject: String,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub paths: Vec<String>,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub commit_sha: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub exit_code: Option<i32>,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub error_class: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub summary: String,
    #[serde(default, skip_serializing_if = "is_zero_timestamp")]
    pub occurred_at: Timestamp,
}

#[derive(Clone, Copy, Debug, Default)]
struct Mapping {
    kind: EventKind,
    phase: EventPhase,
    outcome: EventOutcome,
    error_class: &'static str,
    notification: EventNotificationKind,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct AdapterError(&'static str);

impl fmt::Display for AdapterError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str(self.0)
    }
}

impl std::error::Error for AdapterError {}

#[must_use]
pub fn supported_event_providers() -> Vec<&'static str> {
    vec!["agy", "claude", "codex", "gemini", "kimi", "opencode"]
}

/// Maps a provider-specific event into the closed observation schema.
///
/// # Errors
///
/// Returns a fixed error for unknown providers, types, or disallowed fields.
pub fn normalize_provider_event(
    provider: &str,
    input: ProviderEvent,
) -> Result<EventObservation, AdapterError> {
    let provider = provider.trim().to_ascii_lowercase();
    if provider.is_empty() || provider.len() > 64 || !stable_event_name(&provider) {
        return Err(AdapterError("invalid agent event provider"));
    }
    if input.model_version != PROVIDER_EVENT_MODEL_VERSION || input.sequence == 0 {
        return Err(AdapterError("unsupported provider event contract"));
    }
    let event_type = input.event_type.trim();
    if event_type.is_empty() || event_type.len() > 64 || !stable_event_name(event_type) {
        return Err(AdapterError("invalid provider event type"));
    }
    let lower_type = event_type.to_ascii_lowercase();
    let known_provider = supported_event_providers().contains(&provider.as_str());
    let mut mapped = provider_mapping(&provider, &lower_type);
    if mapped.is_none() && provider == "codex" {
        mapped = codex_item(&lower_type, input.category);
    }
    if mapped.is_none() {
        mapped = canonical_event(&lower_type);
        if let Some(mapping) = mapped {
            if mapping.kind == EventKind::Summary {
                return Err(AdapterError(
                    "provider summaries require an explicit trusted adapter",
                ));
            }
            if !known_provider && mapping.kind != EventKind::Lifecycle {
                return Err(AdapterError("future provider adapter is lifecycle-only"));
            }
        }
    }
    if input.category != EventKind::Unset
        && !(provider == "codex" && lower_type.starts_with("item."))
    {
        return Err(AdapterError(
            "provider event category is not allowed for this type",
        ));
    }
    let Some(mut mapping) = mapped else {
        return Err(AdapterError("unsupported provider event type"));
    };
    if input.exit_code.is_some_and(|code| code != 0)
        && matches!(
            mapping.kind,
            EventKind::Lifecycle | EventKind::Command | EventKind::Test
        )
    {
        mapping.phase = EventPhase::Failed;
        mapping.outcome = EventOutcome::Failed;
        if mapping.notification != EventNotificationKind::Unset {
            mapping.notification = EventNotificationKind::Failure;
        }
    }
    if mapping.notification != EventNotificationKind::Unset {
        return Ok(EventObservation {
            model_version: EVENT_MODEL_VERSION,
            source_id: input.id,
            source_sequence: input.sequence,
            kind: mapping.kind,
            phase: mapping.phase,
            outcome: mapping.outcome,
            error_class: mapping.error_class.to_owned(),
            occurred_at: input.occurred_at,
            notification: mapping.notification,
            recognized_notification: true,
            ..EventObservation::default()
        });
    }
    Ok(EventObservation {
        model_version: EVENT_MODEL_VERSION,
        source_id: input.id,
        source_sequence: input.sequence,
        kind: mapping.kind,
        phase: mapping.phase,
        outcome: mapping.outcome,
        subject: input.subject,
        paths: input.paths,
        commit_sha: input.commit_sha,
        exit_code: input.exit_code,
        error_class: if input.error_class.is_empty() {
            mapping.error_class.to_owned()
        } else {
            input.error_class
        },
        summary: input.summary,
        occurred_at: input.occurred_at,
        ..EventObservation::default()
    })
}

#[allow(clippy::match_same_arms, clippy::too_many_lines)]
fn provider_mapping(provider: &str, event_type: &str) -> Option<Mapping> {
    use EventKind::{Command, Error, File, Lifecycle, Tool};
    use EventNotificationKind::{ApprovalRequested, Completion, Failure, Question};
    use EventOutcome::{Failed, Succeeded};
    use EventPhase::{Completed, Failed as PhaseFailed, Progress, Started, Waiting};
    let value = match (provider, event_type) {
        ("codex" | "claude" | "kimi", "sessionstart") => Mapping {
            kind: Lifecycle,
            phase: Started,
            ..Mapping::default()
        },
        ("codex", "turn.started") | ("kimi", "turnstarted") => Mapping {
            kind: Lifecycle,
            phase: Progress,
            ..Mapping::default()
        },
        ("codex", "turn.completed") | ("codex" | "claude" | "kimi", "stop") => Mapping {
            kind: Lifecycle,
            phase: Waiting,
            ..Mapping::default()
        },
        ("codex", "turn.failed") | ("kimi", "stopfailure") => Mapping {
            kind: Error,
            phase: PhaseFailed,
            outcome: Failed,
            notification: Failure,
            ..Mapping::default()
        },
        ("codex" | "claude" | "kimi", "pretooluse") => Mapping {
            kind: Tool,
            phase: Started,
            ..Mapping::default()
        },
        ("codex" | "claude" | "kimi", "posttooluse") => Mapping {
            kind: Tool,
            phase: Completed,
            outcome: Succeeded,
            ..Mapping::default()
        },
        ("codex" | "claude" | "kimi", "posttoolusefailure") => Mapping {
            kind: Tool,
            phase: PhaseFailed,
            outcome: Failed,
            ..Mapping::default()
        },
        ("codex" | "claude" | "gemini" | "kimi", "permissionrequest") => Mapping {
            kind: Lifecycle,
            phase: Waiting,
            notification: ApprovalRequested,
            ..Mapping::default()
        },
        ("codex" | "claude" | "gemini", "question") => Mapping {
            kind: Lifecycle,
            phase: Waiting,
            notification: Question,
            ..Mapping::default()
        },
        ("codex" | "claude" | "gemini" | "kimi", "sessionend") => Mapping {
            kind: Lifecycle,
            phase: Completed,
            outcome: Succeeded,
            notification: Completion,
            ..Mapping::default()
        },
        ("gemini", "sessionstart") => Mapping {
            kind: Lifecycle,
            phase: Started,
            ..Mapping::default()
        },
        ("gemini", "beforetool") => Mapping {
            kind: Tool,
            phase: Started,
            ..Mapping::default()
        },
        ("gemini", "aftertool") => Mapping {
            kind: Tool,
            phase: Completed,
            outcome: Succeeded,
            ..Mapping::default()
        },
        ("gemini", "notification") => Mapping {
            kind: Lifecycle,
            phase: Progress,
            ..Mapping::default()
        },
        ("agy", "session.started") | ("opencode", "session.created") => Mapping {
            kind: Lifecycle,
            phase: Started,
            ..Mapping::default()
        },
        ("agy", "session.waiting") | ("opencode", "session.idle") => Mapping {
            kind: Lifecycle,
            phase: Waiting,
            ..Mapping::default()
        },
        ("agy" | "opencode", "session.completed") => Mapping {
            kind: Lifecycle,
            phase: Completed,
            outcome: Succeeded,
            notification: Completion,
            ..Mapping::default()
        },
        ("agy", "session.failed") => Mapping {
            kind: Lifecycle,
            phase: PhaseFailed,
            outcome: Failed,
            notification: Failure,
            ..Mapping::default()
        },
        ("agy" | "opencode", "session.question") => Mapping {
            kind: Lifecycle,
            phase: Waiting,
            notification: Question,
            ..Mapping::default()
        },
        ("agy" | "opencode", "permission.requested") => Mapping {
            kind: Lifecycle,
            phase: Waiting,
            notification: ApprovalRequested,
            ..Mapping::default()
        },
        ("opencode", "session.error") => Mapping {
            kind: Error,
            phase: PhaseFailed,
            outcome: Failed,
            error_class: "session_failure",
            notification: Failure,
        },
        ("opencode", "tool.execute.before") => Mapping {
            kind: Tool,
            phase: Started,
            ..Mapping::default()
        },
        ("opencode", "tool.execute.after") => Mapping {
            kind: Tool,
            phase: Completed,
            outcome: Succeeded,
            ..Mapping::default()
        },
        ("opencode", "file.edited") => Mapping {
            kind: File,
            phase: Completed,
            outcome: Succeeded,
            ..Mapping::default()
        },
        ("opencode", "command.executed") => Mapping {
            kind: Command,
            phase: Completed,
            outcome: Succeeded,
            ..Mapping::default()
        },
        _ => return None,
    };
    Some(value)
}

fn codex_item(event_type: &str, category: EventKind) -> Option<Mapping> {
    if !category.is_valid() || matches!(category, EventKind::Lifecycle | EventKind::Summary) {
        return None;
    }
    match event_type {
        "item.started" => Some(Mapping {
            kind: category,
            phase: EventPhase::Started,
            ..Mapping::default()
        }),
        "item.completed" => Some(Mapping {
            kind: category,
            phase: EventPhase::Completed,
            outcome: EventOutcome::Succeeded,
            ..Mapping::default()
        }),
        "item.failed" => Some(Mapping {
            kind: category,
            phase: EventPhase::Failed,
            outcome: EventOutcome::Failed,
            ..Mapping::default()
        }),
        _ => None,
    }
}

fn canonical_event(event_type: &str) -> Option<Mapping> {
    let (kind, phase) = event_type.split_once('.')?;
    if phase.contains('.') {
        return None;
    }
    let kind = parse_kind(kind)?;
    let phase = parse_phase(phase)?;
    let outcome = match phase {
        EventPhase::Completed => EventOutcome::Succeeded,
        EventPhase::Failed => EventOutcome::Failed,
        _ => EventOutcome::Unset,
    };
    let notification = match (kind, phase) {
        (EventKind::Lifecycle, EventPhase::Completed) => EventNotificationKind::Completion,
        (EventKind::Lifecycle, EventPhase::Failed) => EventNotificationKind::Failure,
        _ => EventNotificationKind::Unset,
    };
    Some(Mapping {
        kind,
        phase,
        outcome,
        notification,
        ..Mapping::default()
    })
}

fn parse_kind(value: &str) -> Option<EventKind> {
    Some(match value {
        "lifecycle" => EventKind::Lifecycle,
        "tool" => EventKind::Tool,
        "command" => EventKind::Command,
        "file" => EventKind::File,
        "test" => EventKind::Test,
        "commit" => EventKind::Commit,
        "error" => EventKind::Error,
        "summary" => EventKind::Summary,
        _ => return None,
    })
}

fn parse_phase(value: &str) -> Option<EventPhase> {
    Some(match value {
        "started" => EventPhase::Started,
        "progress" => EventPhase::Progress,
        "waiting" => EventPhase::Waiting,
        "blocked" => EventPhase::Blocked,
        "completed" => EventPhase::Completed,
        "failed" => EventPhase::Failed,
        _ => return None,
    })
}

fn stable_event_name(value: &str) -> bool {
    let mut bytes = value.bytes();
    bytes.next().is_some_and(|byte| byte.is_ascii_alphabetic())
        && bytes.all(|byte| byte.is_ascii_alphanumeric() || matches!(byte, b'.' | b'_' | b'-'))
}

const fn is_unset_kind(value: &EventKind) -> bool {
    matches!(value, EventKind::Unset)
}
const fn is_zero_timestamp(value: &Timestamp) -> bool {
    value.is_zero()
}
